608/708 4 channel support

robwalch opened the pull request:
### This PR will...
1. Add support for 608/708 caption channels 3 and 4
2. Add support for non-native captions track handling
3. Add `attr` and `instreamId` to media playlists (audio and subtitle tracks)

### Why is this Pull Request needed?
This is part of several enhancements that will add parity between the next release (v0.14) and the feature/v1.0.0 branch. These enhancements will help to ensure a cleaner upgrade path to v1.0, by exposing changes in v1.0.0 that were not made in a backwards compatible way (like disabling native captions rendering by default). These changes will get merged into feature/v1.0.0 after release addressing differences in configuration or typing that would make it difficult to switch between versions.

### Are there any points in the code the reviewer needs to double check?

1. Added config options for the new channels. Defaults set to `'Unknown CC'` with empty language code:
```
    captionsTextTrack1Label: 'English', // used by timeline-controller
    captionsTextTrack1LanguageCode: 'en', // used by timeline-controller	    
    captionsTextTrack2Label: 'Spanish', // used by timeline-controller
    captionsTextTrack2LanguageCode: 'es' // used by timeline-controller	    
    captionsTextTrack3Label: 'Unknown CC', // used by timeline-controller
    captionsTextTrack3LanguageCode: '', // used by timeline-controller
    captionsTextTrack4Label: 'Unknown CC', // used by timeline-controller
    captionsTextTrack4LanguageCode: '', // used by timeline-controller
```
2. Demo steam (uses channels 2 and 4? for Spanish and English despite the video saying 1 & 3):
https://deploy-preview-2550--hls-js-pr.netlify.com/demo/?src=%2F%2Fs3.amazonaws.com%2Fqa.jwplayer.com%2F~alex%2FsafarI_captions%2Fnew_master.m3u8&width=720px&config=eyJkZWJ1ZyI6ZmFsc2UsImVuYWJsZVdvcmtlciI6dHJ1ZSwibGl2ZUJhY2tCdWZmZXJMZW5ndGgiOjkwMCwid2lkZXZpbmVMaWNlbnNlVXJsIjoiaHR0cDovL2N3aXAtc2hha2EtcHJveHkuYXBwc3BvdC5jb20vbm9fYXV0aCIsImVtZUVuYWJsZWQiOnRydWUsIm51ZGdlTWF4UmV0cnkiOjV9XBmTVA0IjpmYWxzZSwibGV2ZWxDYXBwaW5nIjotMSwibGltaXRNZXRyaWNzIjotMX0=

### Resolves issues:
Resolves #2329

### Checklist

- [x] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [x] API or design changes are documented in API.md

OrenMe commented on src/controller/timeline-controller.ts:
```diff
@@ -78,7 +88,18 @@ class TimelineController extends EventHandler {
+      const fakeTrack = {
+        cues: [],
+        addCue (cue) {
+          this.cues.pusj(cue);
```
Typo?
```suggestion
          this.cues.push(cue);
```
Didn’t typescript complain about the invalid method call?

robwalch replied:
No because `private Cues: any`. Thanks for the heads up. I'll address the typo and typing.

OrenMe replied:
Didn’t even look at the typings as I assumed it was array of cue object🧐
I just skimmed through the phone for now, I’ll give more thorough look once I get back to the office.

robwalch replied:
@OrenMe I addressed this issue by requiring the cues handler to return an array, and also accept null for the textTrack argument. Since the default behavior is to use native tracks this is backwards compatible. It's also a lot cleaner than what I was doing above even without the typo! https://github.com/video-dev/hls.js/pull/2550/commits/272e1d060215906c9342f5c6a9dd013b15f1ccf6

itsjamie commented on src/controller/timeline-controller.ts:
```diff
@@ -7,24 +7,46 @@ import { logger } from '../utils/logger';
-  private cea608Parser!: Cea608Parser;
+  private unparsedVttFrags: Array<{ frag: Fragment, payload: any }> = [];
+  private cueRanges: Array<[number, number]> = [];
+  private captionsTracks: { [key: string]: TextTrack } = {};
```
I think you could define this as Record<string, TextTrack> to use the embedded TS type for this.

itsjamie commented on src/controller/timeline-controller.ts:
```diff
@@ -329,36 +396,37 @@ class TimelineController extends EventHandler {
-          actualCCBytes.push(ccbyte1);
-          actualCCBytes.push(ccbyte2);
+        if (ccType === 0 || ccType === 1) {
+          actualCCBytes[ccType].push(ccbyte1);
```
Am I correct that this is to add support for ATSC CEA608 as a cctype?

robwalch replied:
This is to store bytes for each channel separately.

itsjamie commented on src/types/media-playlist.ts:
```diff
@@ -1,20 +1,25 @@
-  id: number; // incrementing number to track media playlists
   groupId?: string; // not optional in HLS playlists, but it isn't always specified.
+  id: number; // incrementing number to track media playlists
+  instreamId?: string
```
```suggestion
  instreamId?: string; // ex: "CC1" or "SERVICE2"
```

itsjamie commented on src/controller/timeline-controller.ts:
```diff
@@ -329,36 +396,37 @@ class TimelineController extends EventHandler {
 
     // If the event contains captions (found in the bytes property), push all bytes into the parser immediately
     // It will create the proper timestamps based on the PTS value
+    const cea608Parser = this.cea608Parser as Cea608Parser;
```
Do you need the assertion here? I don't think you do, but haven't checked.

robwalch replied:
Ah. Yeah I think the guard above takes care of that now.

robwalch commented:
@itsjamie Thanks for the feedback. I addressed your comments and did some additional cleanup here https://github.com/video-dev/hls.js/pull/2550/commits/db3fd99a7ce6544b37a9e3386442e5ead663fee7

robwalch commented on src/controller/timeline-controller.ts:
```diff
@@ -200,6 +200,8 @@ class TimelineController extends EventHandler {
     this.prevCC = -1;
     this.vttCCs = newVTTCCs(); // Detect discontinuity in subtitle manifests
     this._cleanTracks();
+    this.tracks = [];
```
This insures we don't get a positive `sameTracks` when reloading a manifest after clearing out the cues. This allows us to properly reuse the tracks when rendering natively but also trigger `NON_NATIVE_TEXT_TRACKS_FOUND` as would be expected after reloading a stream.
